File: CodeChef/November-long-2020/iron-magnet-wall/FEMA2.py
tc = int(input())
while(tc > 0):
    n, k = map(int, input().split())
    s = input()
    p = k+1
    i = 0
    j = 0
    ans = 0
    c = 0
    if(len(s) == 1):
        ans = 0
    else:
        while(i < n and j < n):
            if(s[i] == "M"):
                if(s[j] == "I"):

                    if(i > j):
                        for t in range(j, i+1):
                            if(s[t] == ':'):
                                c = c+1
                    else:
                        for t in range(i, j+1):
                            if(s[t] == ':'):
                                c = c+1

                    f = p-abs(i-j)-c
                    if(f > 0):
                        c = 0
                        ans = ans+1
                        i = i+1
                        j = j+1
                    else:
                        c = 0
                        if(i > j):
                            j = j+1
                        else:
                            i = i+1
                elif(s[j] == "X"):
                    i = j
                    i = i+1
                    j = j+1
                    c = 0
                else:
                    j = j+1
            elif(s[i] == "X"):
                j = i
                c = 0
                i = i+1
                j = j+1

            else:
                i = i+1

    print(ans)

    tc = tc-1

File: CodeChef/November-long-2020/iron-magnet-wall/FEMA2.cpp
//https : //www.codechef.com/NOV20B/problems/FEMA2
#include <bits/stdc++.h>
using namespace std;

int main()
{
    int t;
    int n, k;
    string str;
    cin >> t;
    while (t--)
    {
        int i = 0, j = 0, ans = 0, c = 0, p = 0;
        cin >> n >> k;
        cin >> str;
        p = k + 1;
        while (i < n && j < n)
        {
            if (str[i] == 'M')
            {
                if (str[j] == 'I')
                {
                    if (i > j)
                    {
                        for (int b = j; b < i; b++)
                        {
                            if (str[b] == ':')
                                c++;
                        }
                    }
                    else
                    {
                        for (int b = i; b < j; b++)
                        {
                            if (str[b] == ':')
                                c++;
                        }
                    }
                    p = p - abs(i - j) - c;
                    if (p > 0)
                    {
                        ans++;
                        i++;
                        j++;
                        c = 0;
                    }
                    else
                    {
                        c = 0;
                        if (i > j)
                            j++;
                        else
                            i++;
                    }
                }
                else if (str[j] == 'X')
                {
                    i = j;
                    i++;
                    j++;
                    c = 0;
                }
                else
                {
                    j++;
                }
            }
            else if (str[i] == 'X')
            {
                j = i;
                i++;
                j++;
                c = 0;
            }
            else
            {
                i++;
            }
        }
        cout << ans << endl;
    }
}
